File: fsharp/twelve-days/TwelveDays.fs
module TwelveDays

let recite start stop =
    let gifts = [| 
        "a Partridge in a Pear Tree" 
        "two Turtle Doves";
        "three French Hens";
        "four Calling Birds";
        "five Gold Rings";
        "six Geese-a-Laying";
        "seven Swans-a-Swimming";
        "eight Maids-a-Milking";
        "nine Ladies Dancing";
        "ten Lords-a-Leaping";
        "eleven Pipers Piping";
        "twelve Drummers Drumming";
    |]

    let days = [| "first"; "second"; "third"; "fourth"; "fifth"; "sixth"; 
                  "seventh"; "eighth"; "ninth"; "tenth"; "eleventh"; "twelfth" |]

    let verse n = 
        let currGifts = gifts.[0..n - 1] 
        if n > 1 then currGifts.[0] <- "and " + currGifts.[0]
        let giftsStr = currGifts |> Array.rev |> String.concat ", "
        sprintf "On the %s day of Christmas my true love gave to me: %s." days.[n - 1] giftsStr

    [start..stop] |> List.map verse

File: go/twelve-days/twelve_days.go
package twelve

import (
	"fmt"
	"strings"
)

var (
	verseFmt = "On the %s day of Christmas my true love gave to me: %s."

	gifts = []string{
		"a Partridge in a Pear Tree",
		"two Turtle Doves",
		"three French Hens",
		"four Calling Birds",
		"five Gold Rings",
		"six Geese-a-Laying",
		"seven Swans-a-Swimming",
		"eight Maids-a-Milking",
		"nine Ladies Dancing",
		"ten Lords-a-Leaping",
		"eleven Pipers Piping",
		"twelve Drummers Drumming",
	}

	days = []string{
		"first",
		"second",
		"third",
		"fourth",
		"fifth",
		"sixth",
		"seventh",
		"eighth",
		"ninth",
		"tenth",
		"eleventh",
		"twelfth",
	}
)

// Verse outputs the given N verse of the lyrics to 'The Twelve Days of Christmas'
func Verse(n int) string {
	if n == 1 {
		return fmt.Sprintf(verseFmt, days[0], gifts[0])
	}
	result := make([]string, 0, 12)
	for i := n - 1; i >= 0; i-- {
		result = append(result, gifts[i])
	}
	result[len(result)-1] = "and " + result[len(result)-1]
	return fmt.Sprintf(verseFmt, days[n-1], strings.Join(result, ", "))
}

// Song outputs the lyrics to 'The Twelve Days of Christmas'
func Song() string {
	result := make([]string, 0, 12)
	for i := 1; i <= 12; i++ {
		result = append(result, Verse(i))
	}
	return strings.Join(result, "\n")
}
